polyfill/intersectionobserver-polyfill.js

```diff
@@ -0,0 +1,231 @@
+      }
+      if(this._thresholds instanceof Array) {
+        return this._thresholds;
+      }
+      return [this._thresholds];
+    },
+
+    observe: function(target) {
+      if(!(target instanceof HTMLElement)) {
+        throw('Target needs to be an HTMLelement');
+      }
+
+      var root = this.root;
+      var ancestor = target.parentNode;
+      while (ancestor != root) {
+        if (!ancestor) {
+          throw('Target must be descendant of root');
+        }
+        ancestor = ancestor.parentNode;
+      }
+
+      this._mutationObserver.observe(target, {
+        attributes: true,
+        childList: true,
+        characterData: true,
+        subtree: true
+      });
+      this._observationTargets.set(target, {});
+    },
+
+    unobserve: function(target) {
+      this._observationTargets.delete(target);
+    },
+
+    disconnect: function() {
+      this._observationTargets.clear();
+      this.root.removeEventListener('scroll', this._boundUpdate);
+      this._mutationObserver.disconnect();
+      this._descheduleCallback();
+    },
+
+    takeRecords: function() {
+      this._update();
+      this._descheduleCallback();
+      var copy = this._queue.slice();
+      this._queue = [];
+      return copy;
+    },
+
+    //
+    // Private API
+    //
+    _init: function() {
+      this._observationTargets = new Map();
+      this._boundUpdate = this._update.bind(this);
+      this.root.addEventListener('scroll', debounce(this._boundUpdate, 100));
+      this._mutationObserver = new MutationObserver(debounce(this._boundUpdate, 100));
+      this._queue = [];
+    },
+
+    _update: function() {
+      var rootRect = this._rootRect();
+      this._observationTargets.forEach(function(oldIntersectionEntry, target) {
+        var targetRect = augmentRect(target.getBoundingClientRect());
+        var intersectionRect = intersectRects(rootRect, targetRect);
+        if(!intersectionRect) {
+          return;
+        }
+        var targetArea = targetRect.width * targetRect.height;
+        var intersectionArea = intersectionRect.width * intersectionRect.height;
+        var intersectionRatio = intersectionArea / targetArea;
+        if(!this._hasCrossedThreshold(oldIntersectionEntry.intersectionRatio, intersectionRatio)) {
+          return;
+        }
+        var intersectionEntry = {
+          time: window.performance.now(),
+          rootBounds: rootRect,
+          boundingClientRect: targetRect,
+          intersectionRect: intersectionRect,
+          intersectionRatio: intersectionRatio,
+          target: target
+        };
+        Object.freeze(intersectionEntry);
+        this._queue.push(intersectionEntry);
+        this._scheduleCallback();
+        this._observationTargets.set(target, intersectionEntry);
+      }.bind(this));
+    },
+
+    _scheduleCallback: function() {
+      if(this._idleCallbackId) {
+        return;
+      }
+      this._idleCallbackId = window.requestIdleCallback(function() {
```

Comment:
Whooaaa.. requestIdleCallback is currently an ED that's only implemented in Chrome 47+, right? Can you use a standards track API here like raf or just plain setTimeout?

In the polyfill service we could drop in a simple requestIdleCallback polyfill and depend on that in the IO polyfill, but we prefer not to create dependency relationships on non-standard APIs (we're fine to ship non-standard API polyfills if requested explicitly by the user).  And anyone using this polyfill not via the polyfill service would be stuck.